make a pass at the changes for implementing issue-781 integrating pkg/errors

maleck13 opened the pull request:
@raphael  This is a pass at the changes I proposed for issue-781 https://github.com/goadesign/goa/issues/781

One concern I have here is backwards compatibility. If users are relying on type checks for errors they will stop working.
For example if they were relying on checks like
```
if err, ok := err.(goa.ServiceError); ok {
}

```

From a usability point of view there is no difference however. A user of goa can still do something like

```
return goa.ErrInternal("something strange")

```
and it will work as expected except now you will see a trace in the logs

raphael commented:
Thank you the PR looks great! I was thinking we would let the user call `errors.Wrap` themselves. This way they can provide a meaningful message and only wrap errors they care to wrap. This also helps with the backwards compatibility issue (since the type test is actually documented in the [Error Handling docs](https://goa.design/implement/error_handling/)). There could be a comment in the middleware header explaining that it will take care of logging the callstack if the error is wrapped.

maleck13 commented:
@raphael Thanks for feedback. I will look at updating based on your comment

maleck13 commented:
@raphael Made the changes that I think address your comment. I added a test based on the other error_handler tests and on some of the logging tests. I am not particularly familiar with the test framework so not sure I have implemented the test as would be expected. Happy for any feedback

raphael commented on middleware/error_handler.go:
```diff
@@ -38,7 +40,7 @@ func ErrorHandler(service *goa.Service, verbose bool) goa.Middleware {
 					ctx = context.WithValue(ctx, reqIDKey, reqID)
 				}
-				goa.LogError(ctx, "uncaught error", "id", reqID, "msg", respBody)
+				goa.LogError(ctx, fmt.Sprintf("uncaught error : %+v", e), "id", reqID, "msg", respBody)
```
Is there a reason for not using a key/value pair to log the error? something like `goa.LogError(ctx, "uncaught error", "err", e, "id", reqID, "msg", respBody)`

maleck13 replied:
This seems reasonable, I think it is just a bit of unfamiliarity. Will update PR

raphael commented:
Thank you! the changes look good. I left one comment related to how the logging of uncaught error is done but other than that LGTM.

maleck13 commented:
@raphael do you normally squash commits before merging or are these separate commits ok?

raphael commented:
Yes I'll squash the commits. Is this ready for merge?

maleck13 commented:
👍

raphael commented:
Thank you! do you think you could create a PR against the `v1` branch as well?

maleck13 commented:
@raphael absolutely happy to
